site/src/sphinx/client-circuit-breaker.rst

```diff
@@ -1,6 +1,270 @@
+:apiplural:`Request` exceeds the failure rate. Then the :api:`Client` will immediately get
+:api:`FailFastException` by the :api:`CircuitBreaker`.
+
+.. _circuit-breaker-strategy:
+
+``CircuitBreakerStrategy``
+--------------------------
+
+How does the :api:`CircuitBreaker` know whether a :api:`Response` is successful or not?
+:api:`CircuitBreakerStrategy` does the job. In the above example, if a :api:`Response`'s status is ``5xx``
+or an ``Exception`` is raised during the call, the count of failure is increased.
+Of course, you can have your own ``strategy`` by implementing :api:`CircuitBreakerStrategy`.
+
+.. code-block:: java
+
+    import com.linecorp.armeria.common.HttpStatusClass;
+
+    final CircuitBreakerStrategy<HttpResponse> myStrategy = new CircuitBreakerStrategy<HttpResponse>() {
+
+        @Override
+        public CompletableFuture<Boolean> shouldReportAsSuccess(HttpResponse response) {
+            return response.aggregate().handle((res, cause) -> {
+                if (cause != null) { // A failure if an Exception is raised.
+                    return false;
+                }
+
+                final HttpStatus status = res.status();
+                if (status != null) {
+                    // A failure if the response is 5xx.
+                    if (status.codeClass() == HttpStatusClass.SERVER_ERROR) {
+                        return false;
+                    }
+
+                    // A success if the response is 2xx.
+                    if (status.codeClass() == HttpStatusClass.SUCCESS) {
+                        return true;
+                    }
+                }
+
+                // Not a success nor a failure. Do not include this response.
+                return null;
+            });
+        }
+    };
+
+If you want to treat a :api:`Response` as a success, return ``true``. And return ``false`` as a failure.
+Note that :api:`CircuitBreakerStrategy` can return ``null`` as well. It won't be counted as a success nor
+a failure.
+
+``Grouping``
+------------
+
+In the very first example above, a single :api:`CircuitBreaker` was used to track the events. However,
+using only one :api:`CircuitBreaker` is not recommend. There might be an API which needs heavy calculation
+causing failures frequently. On the other hands, there can be another API which does not need resources
+and simply respond. Having one :api:`CircuitBreaker` that tracks all the success and failure does not make
+sense in this scenario. It's even worse if the :api:`Client` connects to the services on different machines.
+When one of the remote services is down, your :api:`CircuitBreaker` will probably be ``OPEN`` state although
+you can connect to other services.
+Therefore, Armeria provides various ways that let users group the range of circuit breaker instances.
+
+- Per Host: a single :api:`CircuitBreaker` is used for each remote host.
+
+.. code-block:: java
```

Comment:
Could be indented at the same level with the list items? i.e.

```rst
- Per-host: ...

  .. code-block:: java

      import ...
```

Reply:
Oops I didn't know I can do that :)